### PR TITLE
Fix issues with copying worksheets across workbooks

### LINKED ISSUE #829: IXLWorksheet CopyTo does not copy well
**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature

**Version of ClosedXML**

e.g. 0.93.0-beta2

**What is the current behavior?**

When copying a worksheet from a XLTX to a XLSX workbook (it looks like there is the same issue with from a XLSX), the IXLWorksheet's `CopyTo(XLWorkbook workbook, string newSheetName);` method does not:

- keep the row height in the new workbook
- keep the font and font size in all the empty cell

**What is the expected behavior?**

I expect this method to strictly copy the worksheet without changing anything within.

**Did this work in previous versions of our tool?  Which versions?**

Yes, this used to work at least in version 0.88.0 (.NET 4.5.2)

## Reproducibility

**Code to reproduce problem:**
```c#
public void Main()
{
     using (var workbookDocument = new XLWorkbook())
     using (var templateWB = new XLWorkbook(@"Test.xltx"))
     //using (var templateWB = new XLWorkbook(@"Test.xlsx"))
     {
         var ws= templateWB.Worksheet(1);
         ws.CopyTo(workbookDocument, "Copy");
         workbookDocument.SaveAs("Test.xlsx", true);
     }
}
```
- [inputs.zip](https://github.com/ClosedXML/ClosedXML/files/1966487/inputs.zip)

### PR DESCRIPTION
**Pull request by Pankraty**
This PR addresses #829.

Here I added unit tests covering different aspects of worksheets copying and implemented missing functionality.
To simplify the worksheet's method `CopyTo` I extracted part of the logic into corresponding classes (XLTable, XLPicture).

- [ ] C# Code Review: @csreviewer
- [ ] Test Automation Review: @csreviewer

### CODE REVIEW

**Comment by Pankraty** [comment]
Ooops )

File: ClosedXML/Excel/ConditionalFormats/XLConditionalFormat.cs [hunk 1]
```diff
@@ -222,6 +212,19 @@ public IXLConditionalFormat SetStopIfTrue(bool value)
             return this;
         }
 
+        public IXLConditionalFormat CopyTo(IXLWorksheet targetSheet)
+        {
+            var targetRange = targetSheet.Range(Range.RangeAddress.WithoutWorksheet());
```
Comment (igitur) [inline ClosedXML/Excel/ConditionalFormats/XLConditionalFormat.cs]: Do you think that all `CopyTo` methods should first check that `targetSheet` is different to the "current sheet"? For example, it doesn't make sense to copy a conditional format from a sheet into the same sheet.

File: ClosedXML/Excel/Tables/XLTable.cs [hunk 2]
```diff
@@ -817,5 +817,39 @@ public DataTable AsNativeDataTable()
...
+        {
+            return CopyTo((XLWorksheet) targetSheet);
+        }
+
+        internal IXLTable CopyTo(XLWorksheet targetSheet)
+        {
+            String tableName = Name;
+            var newTable = (XLTable)targetSheet.Table(targetSheet.Range(RangeAddress.ToString()), tableName, true);
+
+            newTable.RelId = RelId;
```
Comment (igitur) [inline ClosedXML/Excel/Tables/XLTable.cs]: The `RelId` is to store the `relId` when a workbook is opened. When creating a new table (by copying it), the `RelId` should remain empty.

File: ClosedXML/Excel/Tables/XLTable.cs [hunk 3]
```diff
@@ -817,5 +817,39 @@ public DataTable AsNativeDataTable()
...
         }
+
+        public IXLTable CopyTo(IXLWorksheet targetSheet)
+        {
+            return CopyTo((XLWorksheet) targetSheet);
+        }
+
+        internal IXLTable CopyTo(XLWorksheet targetSheet)
+        {
+            String tableName = Name;
```
Comment (igitur) [inline ClosedXML/Excel/Tables/XLTable.cs]: Should copying a table not also involve copying the contents of the cells in the table, i.e. the range should be copied too?

File: ClosedXML_Tests/Excel/Worksheets/XLWorksheetTests.cs [hunk 4]
```diff
@@ -1,8 +1,11 @@
 using ClosedXML.Excel;
 using NUnit.Framework;
 using System;
+using System.Drawing;
 using System.IO;
 using System.Linq;
+using System.Reflection;
+using ClosedXML.Excel.Drawings;
```
Comment (igitur) [inline ClosedXML_Tests/Excel/Worksheets/XLWorksheetTests.cs]: Alphabetical

File: ClosedXML/Excel/ConditionalFormats/XLConditionalFormat.cs (hunk 1 above)
Comment (Pankraty) [reply ClosedXML/Excel/ConditionalFormats/XLConditionalFormat.cs]: True. Added a validation and two more unit tests.

File: ClosedXML/Excel/ConditionalFormats/XLConditionalFormat.cs [hunk 5]
```diff
@@ -222,6 +212,19 @@ public IXLConditionalFormat SetStopIfTrue(bool value)
...
+            var newCf = new XLConditionalFormat(this, targetRange);
```
[the remark this answers is not in the document]
Comment (Pankraty) [reply ClosedXML/Excel/ConditionalFormats/XLConditionalFormat.cs]: Good idea.

File: ClosedXML/Excel/Drawings/XLPicture.cs [hunk 6]
```diff
@@ -326,6 +326,44 @@ public IXLPicture WithSize(Int32 width, Int32 height)
...
         }
 
+        public IXLPicture CopyTo(IXLWorksheet targetSheet)
+        {
+            return CopyTo((XLWorksheet) targetSheet);
+        }
+
+        internal IXLPicture CopyTo(XLWorksheet targetSheet)
+        {
+            var newPicture = targetSheet.AddPicture(ImageStream, Format, Name)
```
[the remark this answers is not in the document]
Comment (Pankraty) [reply ClosedXML/Excel/Drawings/XLPicture.cs]: This is actually done in `XLPicture` constructor:
```
            this.ImageStream = new MemoryStream();
            {
                stream.Position = 0;
                stream.CopyTo(ImageStream);
                ImageStream.Seek(0, SeekOrigin.Begin);
```

[the remark this answers is not in the document]
Comment (igitur) [reply ClosedXML/Excel/Drawings/XLPicture.cs]: I can't remember my own code! I still feel we should change to byte array. But that's a separate issue.

File: ClosedXML/Excel/Tables/XLTable.cs (hunk 2 above)
Comment (Pankraty) [reply ClosedXML/Excel/Tables/XLTable.cs]: Perhaps you are right. I just moved this from `XLWorksheet` class. And this code was quite ancient (2011, commit message "worksheet.CopyTo(...) should now work as expected." LOL)

Comment (igitur) [reply ClosedXML/Excel/Tables/XLTable.cs]: If it's from 2011, then it's before my time. Let me just double-check my previous statement in this regard first.

File: ClosedXML/Excel/Tables/XLTable.cs (hunk 3 above)
Comment (Pankraty) [reply ClosedXML/Excel/Tables/XLTable.cs]: As long as we use this method in the scope of worksheet copying only it is not needed as cells are copied separately. But if we would like to use this method separately, we'd have to perform cells copying as well. Although I do not see any use cases that would require it.

Comment (igitur) [reply ClosedXML/Excel/Tables/XLTable.cs]: Well, you included it in `IXLTable.cs`, same with the other `CopyTo` methods. So people are going to use it. And I think it's a good idea to have this publicly exposed.

File: ClosedXML_Tests/Excel/Worksheets/XLWorksheetTests.cs [hunk 7]
```diff
@@ -296,5 +300,522 @@ public void WorksheetNameCanContainApostrophe()
...
+                var ws1 = wb1.Worksheets.Add("Original");
+
+                ws1.Cell("A1").FormulaA1 = "10*10";
+                ws1.Cell("A2").FormulaA1 = "Original!A1 * 3";
+
+                ws1.Name = "Renamed";
+
+                Assert.AreEqual("Renamed!A1 * 3", ws1.Cell("A2").FormulaA1);
+            }
+        }
```
Comment (igitur) [inline ClosedXML_Tests/Excel/Worksheets/XLWorksheetTests.cs]: Also add individual tests for e.g. `IXLTable.CopyTo` in the file for testing tables. Same with the other objects.

File: ClosedXML/Excel/Tables/XLTable.cs (hunk 2 above)
Comment (Pankraty) [reply ClosedXML/Excel/Tables/XLTable.cs]: I'll add another test tomorrow.

Comment (Pankraty) [reply ClosedXML/Excel/Tables/XLTable.cs]: When I remove this line `CopyWorksheets` test fails. The produced file is correct either way, just table IDs differ.

File: ClosedXML_Tests/Excel/Worksheets/XLWorksheetTests.cs (hunk 7 above)
Comment (Pankraty) [reply ClosedXML_Tests/Excel/Worksheets/XLWorksheetTests.cs]: Test for tables, pictures and named ranges added. Conditional formats already had them.

**Comment by Pankraty** [comment]
The first set of feedback processed.

File: ClosedXML/Excel/Drawings/IXLPicture.cs [hunk 8]
```diff
@@ -8,6 +8,14 @@ public interface IXLPicture : IDisposable
     {
         IXLAddress BottomRightCellAddress { get; }
 
+        /// <summary>
+        /// Create a copy of the picture.
+        /// </summary>
+        /// <param name="targetSheet">A worksheet where put the copy to. When null the copy of the picture is created
```
Comment (igitur) [inline ClosedXML/Excel/Drawings/IXLPicture.cs]: I don't think we should allow `targetSheet == null`. I would rather have a separate `IXLPicture.Duplicate()` method, which does the same.

Also update the parameter doctext to `The worksheet to which the picture will be copied.`

File: ClosedXML/Excel/Drawings/IXLPicture.cs [hunk 9]
```diff
@@ -8,6 +8,19 @@ public interface IXLPicture : IDisposable
     {
         IXLAddress BottomRightCellAddress { get; }
 
+        /// <summary>
+        /// Create a copy of the picture on a different worksheet.
+        /// </summary>
+        /// <param name="targetSheet">The worksheet to which the picture will be copied.</param>
+        /// <returns>A created copy of the picture.</returns>
+        IXLPicture CopyTo(IXLWorksheet targetSheet = null);
```
Comment (igitur) [inline ClosedXML/Excel/Drawings/IXLPicture.cs]: Make the parameter non-optional, please.

**Comment by igitur** [comment]
Great work, again.

**Comment by Pankraty** [comment]
Thanks )

**Pull request merged**